### PR TITLE
Add convenience routines for DP-SGD

### PR DESCRIPTION
**Pull request by wulu473**
We're adding some useful functions and utilities specifically for DP-SGD training.
- DPSGD Accountant which is short hand for the poisson subsampled Gaussian mechanism
- `find_noise_multiplier` which uses the `DPSGDAccountant` to give an appropriate `noise_multiplier` given a desired target epsilon.

### CODE REVIEW

File: prv_accountant/dpsgd.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    :param float target_delta: Value of DP delta
+    :param float eps_error: Error allowed for final epsilon
+    """
+    def compute_epsilon(mu: float) -> float:
+        acc = DPSGDAccountant(
+            noise_multiplier=mu,
+            sampling_probability=sampling_probability,
+            max_steps=num_steps,
+            eps_error=eps_error/2,
+            delta_error=target_delta/1000
```
Comment (donebydan) [inline prv_accountant/dpsgd.py:56]: Should the delta_error be dependent on number of samples here and not just `target_delta / 1000`? Something like `target_delta / f(#samples)`?

File: prv_accountant/dpsgd.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        )
+        return acc.compute_epsilon(delta=target_delta, num_steps=num_steps)
+
+    mu_max = 100.0
```
Comment (donebydan) [inline prv_accountant/dpsgd.py:60]: Could `mu_max` be a variable for streamlined use? Say you don't care for `mu_max > 50`.

File: tests/test_dpsgd.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class TestFindNoiseMultiplier:
+    def test_sensible_range(self):
+        mu = find_noise_multiplier(sampling_probability=2e-3, num_steps=10_000, target_epsilon=4.0, target_delta=1e-7)
+        assert 0 < mu and mu < 2  # Check that mu is in a sensible interval
+
+    def test_inverse(self):
+        mu = find_noise_multiplier(2e-3, 10_000, 4.0, 1e-7)
+        acc = DPSGDAccountant(mu, 2e-3, max_steps=10_000, eps_error=0.5)
+        eps = acc.compute_epsilon(delta=1e-7, num_steps=10_000)
```
[the remark this answers is not in the document]
Comment (donebydan) [reply tests/test_dpsgd.py:14]: Maybe even test class variables since they are common?

File: tests/test_dpsgd.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                sampling_probability=256/50_000,
+                num_steps=int(50*50_000/256),
+                target_epsilon=10.0,
+                target_delta=1e-5
+            )
+        assert len(record) == 0
+
+    def test_robustness_2(self):
+        mu = find_noise_multiplier(
+            sampling_probability=0.26058631921824105,
```
Comment (donebydan) [inline tests/test_dpsgd.py:29]: NIT: should we be consistent how we pass these variables. For example, `sampling_probability` is a fraction, float and scientific notations in this test class.

File: tests/test_dpsgd.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            num_steps=18800,
+            target_delta=0.00011448277499759097,
+            target_epsilon=4.0
+        )
+        # Just test that this doesn't cause a floating point overflow
```
Comment (donebydan) [inline tests/test_dpsgd.py:34]: I would prefer code to this comment. Perhaps we could actually assert this by specifically failing the test if it does cause floating point (it won't). Or some alternative.

File: prv_accountant/dpsgd.py (hunk 1 above)
Comment (wulu473) [reply prv_accountant/dpsgd.py:56]: target delta should already take into account the number of samples.

File: prv_accountant/dpsgd.py (hunk 2 above)
Comment (wulu473) [reply prv_accountant/dpsgd.py]: Good point. I've made it a variable.

File: tests/test_dpsgd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wulu473) [reply tests/test_dpsgd.py]: Done.

File: tests/test_dpsgd.py (hunk 5 above)
Comment (wulu473) [reply tests/test_dpsgd.py]: Done.

File: tests/test_dpsgd.py (hunk 4 above)
Comment (wulu473) [reply tests/test_dpsgd.py]: These are specific edge cases that might be worthwhile to keep exactly in the form as they were reported.

**Review by donebydan — APPROVED** [review]
